### PR TITLE
Use tachyon to speed up witness and proofgen #75

### LINKED ISSUE #75: Use tachyon to speed up witness and proofgen
https://github.com/kroma-network/tachyon

Make sure to update our Docker files and benchmark it against our current rapid snark

### PR DESCRIPTION
**Pull request by whoisgautxm**
Fixes #75 

This PR integrates Tachyon to speed up the witness and proof generation process. The following changes were made:

Changed libs/rapidSnark.Dockerfile to libs/tachyon.Dockerfile
Changed the packages/prover/circom_proofgen.sh to accommodate proofgen by tachyon
Changes the packages/prover/Dockerfile to accommodate tachyon prover
Changes the packages/prover/core.py to accommodate tachyon prover
The benchmark results of circom circuits are in this github repo: https://github.com/whoisgautxm/circom-benchmark, which show ~30% speedup due to nmap/mmap usage primarily.

### CODE REVIEW

**Comment by SoraSuegami** [comment]
Thank you for your PR!
I am sorry for the delay in reviewing it.
When I ran `modal deploy modal_server.py`, I got the following error when building a docker image.
```
=> Step 24: RUN git fetch origin && git checkout build/support-building-tachyon-based-on-python-3.10-docker-image
error: pathspec 'build/support-building-tachyon-based-on-python-3.10-docker-image' did not match any file(s) known to git
Terminating task due to error: failed to run builder command "git fetch origin && git checkout build/support-building-tachyon-based-on-python-3.10-docker-image"
```

I think the Dockerfile should specify main branch of tachyon, right?

**Comment by SoraSuegami** [comment]
I changed the branch to main, and got the following error.
```
ERROR: /tachyon/tachyon/zk/air/plonky3/challenger/BUILD.bazel:53:20: Compiling tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc failed: (Exit 1): gcc failed: error executing command (from target //tachyon/zk/air/plonky3/challenger:challenger_unittests) /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 97 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from external/com_google_absl/absl/container/inlined_vector.h:53,
                 from ./tachyon/zk/air/plonky3/challenger/duplex_challenger.h:11,
                 from tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc:6:
In member function 'void absl::lts_20230125::inlined_vector_internal::Storage<T, N, A>::MemcpyFrom(const absl::lts_20230125::inlined_vector_internal::Storage<T, N, A>&) [with T = tachyon::math::PrimeField<tachyon::math::BabyBearConfig>; long unsigned int N = 4; A = std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> >]',
    inlined from 'void absl::lts_20230125::InlinedVector<T, N, A>::MoveAssignment(MemcpyPolicy, absl::lts_20230125::InlinedVector<T, N, A>&&) [with T = tachyon::math::PrimeField<tachyon::math::BabyBearConfig>; long unsigned int N = 4; A = std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> >]' at external/com_google_absl/absl/container/inlined_vector.h:790:24,
    inlined from 'absl::lts_20230125::InlinedVector<T, N, A>& absl::lts_20230125::InlinedVector<T, N, A>::operator=(absl::lts_20230125::InlinedVector<T, N, A>&&) [with T = tachyon::math::PrimeField<tachyon::math::BabyBearConfig>; long unsigned int N = 4; A = std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> >]' at external/com_google_absl/absl/container/inlined_vector.h:498:21,
    inlined from 'tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > > >, 16, 4>& tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > > >, 16, 4>::operator=(tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > > >, 16, 4>&&)' at ./tachyon/zk/air/plonky3/challenger/duplex_challenger.h:19:7,
    inlined from 'virtual void tachyon::zk::air::plonky3::{anonymous}::DuplexChallengerTest::SetUp()' at tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc:34:79:
external/com_google_absl/absl/container/internal/inlined_vector.h:468:5: error: '<unnamed>.tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig, void> > > >, 16, 4>::input_buffer_.absl::lts_20230125::InlinedVector<tachyon::math::PrimeField<tachyon::math::BabyBearConfig, void>, 4, std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig, void> > >::storage_.absl::lts_20230125::inlined_vector_internal::Storage<tachyon::math::PrimeField<tachyon::math::BabyBearConfig>, 4, std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > >::data_' may be used uninitialized [-Werror=maybe-uninitialized]
  468 |     data_ = other_storage.data_;
      |     ^~~~~
tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc: In member function 'virtual void tachyon::zk::air::plonky3::{anonymous}::DuplexChallengerTest::SetUp()':
tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc:34:79: note: '<anonymous>' declared here
   34 |     challenger_ = DuplexChallenger<Poseidon2, kWidth, kRate>(std::move(sponge));
      |                                                                               ^
In member function 'void absl::lts_20230125::inlined_vector_internal::Storage<T, N, A>::MemcpyFrom(const absl::lts_20230125::inlined_vector_internal::Storage<T, N, A>&) [with T = tachyon::math::PrimeField<tachyon::math::BabyBearConfig>; long unsigned int N = 16; A = std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> >]',
    inlined from 'void absl::lts_20230125::InlinedVector<T, N, A>::MoveAssignment(MemcpyPolicy, absl::lts_20230125::InlinedVector<T, N, A>&&) [with T = tachyon::math::PrimeField<tachyon::math::BabyBearConfig>; long unsigned int N = 16; A = std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> >]' at external/com_google_absl/absl/container/inlined_vector.h:790:24,
    inlined from 'absl::lts_20230125::InlinedVector<T, N, A>& absl::lts_20230125::InlinedVector<T, N, A>::operator=(absl::lts_20230125::InlinedVector<T, N, A>&&) [with T = tachyon::math::PrimeField<tachyon::math::BabyBearConfig>; long unsigned int N = 16; A = std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> >]' at external/com_google_absl/absl/container/inlined_vector.h:498:21,
    inlined from 'tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > > >, 16, 4>& tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > > >, 16, 4>::operator=(tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > > >, 16, 4>&&)' at ./tachyon/zk/air/plonky3/challenger/duplex_challenger.h:19:7,
    inlined from 'virtual void tachyon::zk::air::plonky3::{anonymous}::DuplexChallengerTest::SetUp()' at tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc:34:79:
external/com_google_absl/absl/container/internal/inlined_vector.h:468:5: error: '<unnamed>.tachyon::zk::air::plonky3::DuplexChallenger<tachyon::crypto::Poseidon2Sponge<tachyon::crypto::Poseidon2ExternalMatrix<tachyon::crypto::Poseidon2Plonky3ExternalMatrix<tachyon::math::PrimeField<tachyon::math::BabyBearConfig, void> > > >, 16, 4>::output_buffer_.absl::lts_20230125::InlinedVector<tachyon::math::PrimeField<tachyon::math::BabyBearConfig, void>, 16, std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig, void> > >::storage_.absl::lts_20230125::inlined_vector_internal::Storage<tachyon::math::PrimeField<tachyon::math::BabyBearConfig>, 16, std::allocator<tachyon::math::PrimeField<tachyon::math::BabyBearConfig> > >::data_' may be used uninitialized [-Werror=maybe-uninitialized]
  468 |     data_ = other_storage.data_;
      |     ^~~~~
tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc: In member function 'virtual void tachyon::zk::air::plonky3::{anonymous}::DuplexChallengerTest::SetUp()':
tachyon/zk/air/plonky3/challenger/duplex_challenger_unittest.cc:34:79: note: '<anonymous>' declared here
   34 |     challenger_ = DuplexChallenger<Poseidon2, kWidth, kRate>(std::move(sponge));
      |                                                                               ^
cc1plus: all warnings being treated as errors
INFO: Elapsed time: 502.849s, Critical Path: 147.82s
INFO: 4948 processes: 1687 internal, 396 local, 2865 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
Terminating task due to error: failed to run builder command "bazel build //..."
```

**Comment by SoraSuegami** [comment]
Thank you for your additional comments!
I reflected changes you suggested, and I next got the following error.
```
=> Step 24: RUN CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index && bazel build --@kroma_network_tachyon//:has_openmp -c opt --config linux //:prover_main
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
 checking cached actions
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/kroma-network/ffiasm/archive/7db5a442b12dba825ec35361441aa116dc3bff8b.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
 checking cached actions
 checking cached actions
 checking cached actions
Loading: 
Loading: 
Loading: 
Loading: 1 packages loaded
Analyzing: target //:prover_main (2 packages loaded, 0 targets configured)
Analyzing: target //:prover_main (88 packages loaded, 223 targets configured)
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/ingonyama-zk/icicle/archive/4fef5423462a72a37fe66cee89e24eae083cc112.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
Analyzing: target //:prover_main (151 packages loaded, 668 targets configured)
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/orlp/pdqsort/archive/b1ef26a55cdb60d236a5cb199c4234c704f46726.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
DEBUG: Rule 'com_github_gflags_gflags' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "34af2f15cf7367513b352bdcd2493ab14ce43692d2dcd9dfc499492966c64dcf"
DEBUG: Repository com_github_gflags_gflags instantiated at:
  /tachyon/vendors/circom/WORKSPACE:56:13: in <toplevel>
  /root/.cache/bazel/_bazel_root/0f899c535a62c364a7f75378d7f046ec/external/kroma_network_tachyon/bazel/tachyon_deps.bzl:65:21: in tachyon_deps
Repository rule http_archive defined at:
  /root/.cache/bazel/_bazel_root/0f899c535a62c364a7f75378d7f046ec/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
WARNING: Download from https://mirror.bazel.build/github.com/gflags/gflags/archive/v2.2.2.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/sebawild/powersort/archive/48e31e909280ca43bb2c33dd3df9922b0a0f3f84.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
Analyzing: target //:prover_main (165 packages loaded, 1430 targets configured)
Analyzing: target //:prover_main (171 packages loaded, 1845 targets configured)
Analyzing: target //:prover_main (179 packages loaded, 7332 targets configured)
Analyzing: target //:prover_main (349 packages loaded, 8572 targets configured)
Analyzing: target //:prover_main (350 packages loaded, 8647 targets configured)
Analyzing: target //:prover_main (350 packages loaded, 8647 targets configured)
INFO: Analyzed target //:prover_main (351 packages loaded, 8673 targets configured).
INFO: Found 1 target...
[0 / 308] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[492 / 1,043] Compiling absl/base/internal/throw_delegate.cc; 0s processwrapper-sandbox ... (51 actions, 48 running)
[645 / 1,114] Compiling absl/numeric/int128.cc; 1s processwrapper-sandbox ... (53 actions, 51 running)
[824 / 1,208] Compiling absl/strings/internal/cord_rep_btree_navigator.cc; 2s processwrapper-sandbox ... (52 actions, 50 running)
[932 / 1,259] Compiling absl/strings/internal/cord_internal.cc; 3s processwrapper-sandbox ... (51 actions, 49 running)
[1,017 / 1,328] Compiling absl/strings/internal/cordz_info.cc; 4s processwrapper-sandbox ... (51 actions, 50 running)
[1,039 / 1,330] Compiling absl/strings/cord.cc; 5s processwrapper-sandbox ... (58 actions, 57 running)
[1,056 / 1,330] Compiling absl/strings/cord.cc; 6s processwrapper-sandbox ... (58 actions, 57 running)
[1,075 / 1,330] Compiling absl/strings/cord.cc; 7s processwrapper-sandbox ... (58 actions, 57 running)
[1,086 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 8s local ... (58 actions, 57 running)
[1,101 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 9s local ... (57 actions, 56 running)
[1,106 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 11s local ... (57 actions running)
[1,124 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 12s local ... (58 actions, 57 running)
[1,146 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 13s local ... (58 actions, 57 running)
[1,175 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 14s local ... (58 actions, 57 running)
[1,208 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 15s local ... (53 actions running)
[1,250 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 16s local ... (12 actions running)
[1,253 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 17s local ... (11 actions running)
[1,262 / 1,330] Creating runfiles tree bazel-out/k8-opt-exec-2B5CBBC6/bin/external/aspect_rules_js/npm/private/lifecycle/lifecycle-hooks.sh.runfiles [for tool]; 18s local ... (6 actions running)
[1,281 / 1,330] Compiling sdk/perfetto.cc; 12s processwrapper-sandbox ... (12 actions, 5 running)
[1,312 / 1,330] Compiling sdk/perfetto.cc; 14s processwrapper-sandbox ... (13 actions running)
[1,317 / 1,330] Compiling sdk/perfetto.cc; 16s processwrapper-sandbox ... (8 actions running)
[1,318 / 1,330] Compiling sdk/perfetto.cc; 19s processwrapper-sandbox ... (7 actions running)
[1,319 / 1,330] Compiling sdk/perfetto.cc; 23s processwrapper-sandbox ... (6 actions running)
[1,323 / 1,330] Compiling sdk/perfetto.cc; 24s processwrapper-sandbox ... (6 actions, 5 running)
[1,327 / 1,330] Compiling sdk/perfetto.cc; 26s processwrapper-sandbox ... (2 actions running)
ERROR: /tachyon/vendors/circom/BUILD.bazel:3:18: Compiling prover_main.cc failed: (Exit 1): gcc failed: error executing command (from target //:prover_main) /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 123 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from external/kroma_network_tachyon/tachyon/math/polynomials/univariate/univariate_evaluation_domain_factory.h:11,
                 from prover_main.cc:16:
In static member function 'static constexpr std::unique_ptr<tachyon::math::MixedRadixEvaluationDomain<F, MaxDegree> > tachyon::math::MixedRadixEvaluationDomain<F, MaxDegree>::Create(size_t) [with F = tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig>; long unsigned int MaxDegree = 18446744073709551615]',
    inlined from 'static constexpr std::unique_ptr<tachyon::math::UnivariateEvaluationDomain<F, MaxDegree> > tachyon::math::UnivariateEvaluationDomainFactory<F, MaxDegree>::Create(size_t) [with F = tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig>; long unsigned int MaxDegree = 18446744073709551615]' at external/kroma_network_tachyon/tachyon/math/polynomials/univariate/univariate_evaluation_domain_factory.h:39:73,
    inlined from 'static constexpr std::unique_ptr<tachyon::math::UnivariateEvaluationDomain<F, MaxDegree> > tachyon::math::UnivariateEvaluationDomain<F, MaxDegree>::Create(size_t) [with F = tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig>; long unsigned int MaxDegree = 18446744073709551615]' at external/kroma_network_tachyon/tachyon/math/polynomials/univariate/univariate_evaluation_domain.h:75:78,
    inlined from 'void tachyon::circom::CreateProof(const tachyon::base::FilePath&, const tachyon::base::FilePath&, const tachyon::base::FilePath&, const tachyon::base::FilePath&, const CreateProofOptions&) [with Curve = tachyon::math::BLS12Curve<tachyon::math::bls12_381::BLS12_381Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig>, tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > >, tachyon::math::Fp6<tachyon::math::bls12_381::Fq6Config<tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > > > >, tachyon::math::Fp12<tachyon::math::bls12_381::Fq12Config<tachyon::math::Fp6<tachyon::math::bls12_381::Fq6Config<tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > > > > > >, tachyon::math::SWCurve<tachyon::math::bls12_381::G1CurveConfig<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig>, tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig> > >, tachyon::math::SWCurve<tachyon::math::bls12_381::G2CurveConfig<tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > >, tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig> > > > >]' at prover_main.cc:116:27:
external/kroma_network_tachyon/tachyon/math/polynomials/univariate/mixed_radix_evaluation_domain.h:70:9: error: 'factors.tachyon::math::PrimeFieldFactors::two_adicity' may be used uninitialized [-Werror=maybe-uninitialized]
   70 |         new MixedRadixEvaluationDomain(size, factors.two_adicity));
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
external/kroma_network_tachyon/tachyon/math/polynomials/univariate/mixed_radix_evaluation_domain.h: In function 'void tachyon::circom::CreateProof(const tachyon::base::FilePath&, const tachyon::base::FilePath&, const tachyon::base::FilePath&, const tachyon::base::FilePath&, const CreateProofOptions&) [with Curve = tachyon::math::BLS12Curve<tachyon::math::bls12_381::BLS12_381Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig>, tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > >, tachyon::math::Fp6<tachyon::math::bls12_381::Fq6Config<tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > > > >, tachyon::math::Fp12<tachyon::math::bls12_381::Fq12Config<tachyon::math::Fp6<tachyon::math::bls12_381::Fq6Config<tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > > > > > >, tachyon::math::SWCurve<tachyon::math::bls12_381::G1CurveConfig<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig>, tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig> > >, tachyon::math::SWCurve<tachyon::math::bls12_381::G2CurveConfig<tachyon::math::Fp2<tachyon::math::bls12_381::Fq2Config<tachyon::math::PrimeField<tachyon::math::bls12_381::FqConfig> > >, tachyon::math::PrimeField<tachyon::math::bls12_381::FrConfig> > > > >]':
external/kroma_network_tachyon/tachyon/math/polynomials/univariate/mixed_radix_evaluation_domain.h:67:23: note: 'factors' declared here
   67 |     PrimeFieldFactors factors;
      |                       ^~~~~~~
cc1plus: all warnings being treated as errors
Target //:prover_main failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 56.433s, Critical Path: 38.74s
INFO: 1277 processes: 564 internal, 396 local, 317 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
Terminating task due to error: failed to run builder command "CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index && bazel build --@kroma_network_tachyon//:has_openmp -c opt --config linux //:prover_main"

Caused by:
    container exit status: 1
```

File: packages/prover/Dockerfile [hunk 1]
```diff
@@ -1,52 +1,72 @@
...
+RUN curl -fsSL https://bazel.build/bazel-release.pub.gpg | gpg --dearmor > bazel.gpg && \
+  mv bazel.gpg /etc/apt/trusted.gpg.d/ && \
+  echo "deb [arch=amd64] https://storage.googleapis.com/bazel-apt stable jdk1.8" | tee /etc/apt/sources.list.d/bazel.list && \
+  apt-get update && \
+  apt-get install -y bazel-6.3.0 && \
+  ln -s /usr/bin/bazel-6.3.0 /usr/local/bin/bazel
+
+
+# Clone and build Tachyon
+RUN git clone https://github.com/kroma-network/tachyon /tachyon
```
[the remark this answers is not in the document]
Comment (chokobole) [reply packages/prover/Dockerfile:64]: @SoraSuegami I think perhaps you are missing this!

[the remark this answers is not in the document]
Comment (chokobole) [reply packages/prover/Dockerfile:64]: @SoraSuegami I think perhaps you are missing this!

[the remark this answers is not in the document]
Comment (SoraSuegami) [reply packages/prover/Dockerfile]: Thank you very much!
I could build the docker image on modal!

**Comment by chokobole** [comment]
We don't store points in a vector of strings, as demonstrated in the code [here](https://github.com/kroma-network/tachyon/blob/main/tachyon/zk/r1cs/groth16/proof.h). However, the output of prover_main is identical to that of a proof generated by rapidsnark. You can see more details in this [file](https://github.com/kroma-network/tachyon/blob/main/vendors/circom/circomlib/json/groth16_proof.h).

**Comment by Divide-By-0** [comment]
If our parsing is wrong, are we confident that a new modal deployment will still be able to generate the correct format of proof? Was this end-to-end tested at all

**Comment by SoraSuegami** [comment]
@chokobole 
I got the following json from a prover.
```
"proof": [
    "2018721414038404820327",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "6632353713085157925504008443078919716322386156160602218536961028046468237192",
    "19894320019016719114298695954688611116822060474192999674641264748344728553214",
    "1724342615",
    "12424312113957298897351547719054010854075115739442060206737756826328039462065",
    "4294223308299735140139693782429412250667096998968531391722386702932610034429",
    "10287587543879982271389500964794028174629429687291862342493089880334707793270"
  ]
```

I am not sure why it is different from the expected format in the code.   
Could you tell me how to fix that, or how can the relayer decompose it into pi_a, pi_b, and pi_c parts?

**Comment by SoraSuegami** [comment]
> If our parsing is wrong, are we confident that a new modal deployment will still be able to generate the correct format of proof? Was this end-to-end tested at all

Sorry, I merged this PR because the modal sever pulls this repo from main branch when building a Dockerfile, and I expected that the prover worked once a docker image is built.

**Comment by chokobole** [comment]
@SoraSuegami Could you tell me how to run the tests?

**Comment by Bisht13** [comment]
@chokobole the tests still seems to be slower than rapidsnark, maybe we can see some sort of acceleration by using gpu?

**Pull request merged**